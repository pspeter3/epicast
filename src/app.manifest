{
    "background_color": "#5b8121",
    "dir": "ltr",
    "display": "standalone",
    "icons": [
        {
            "src": "/epidemia/logo_192.02d693d5.png",
            "type": "image/png",
            "sizes": "192x192"
        },
        {
            "src": "/epidemia/logo_512.3ef7cff0.png",
            "type": "image/png",
            "sizes": "512x512"
        }
    ],
    "lang": "en",
    "name": "Epidemia: Pandemic Forecaster",
    "orientation": "portrait",
    "scope": "/epidemia/",
    "short_name": "Epidemia",
    "start_url": "/epidemia/#/",
    "theme_color": "#272727"
}
