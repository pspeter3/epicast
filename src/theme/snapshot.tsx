import * as React from "react";
import TestRenderer from "react-test-renderer";

export const snapshot = (name: string, SFC: React.SFC<{}>): void => {
    describe(name, () => {
        it(`should match snapshot`, () => {
            const component = TestRenderer.create(<SFC>Snapshot</SFC>);
            const tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });
};
