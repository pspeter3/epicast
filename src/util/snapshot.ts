import { shallow } from "enzyme";
import * as React from "react";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        expect(shallow(factory())).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
