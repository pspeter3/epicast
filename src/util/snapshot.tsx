import * as React from "react";
import TestRenderer from "react-test-renderer";

export type Factory = () => React.ReactElement<any>;

export const snapshot = (factory: Factory): (() => void) => {
    return () => {
        const component = TestRenderer.create(factory());
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    };
};

export const snapshotSuite = (name: string, factory: Factory): void => {
    describe(name, () => {
        it("should match snapshot", snapshot(factory));
    });
};
