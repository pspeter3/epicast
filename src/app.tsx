import * as React from "react";
import {
    AppBar,
    AppBarIcon,
    AppBarNav,
    AppBarSection,
    AppBarTab,
    AppBarTitle,
} from "./theme/appbar";
import { SettingsIcon } from "./theme/icons";

export class App extends React.Component<{}, {}> {
    public render() {
        return (
            <AppBar>
                <AppBarSection>
                    <AppBarTitle>Epidemia</AppBarTitle>
                    <AppBarIcon to="/settings">
                        <SettingsIcon />
                    </AppBarIcon>
                </AppBarSection>
                <AppBarNav>
                    <AppBarTab to="/">Dashboard</AppBarTab>
                    <AppBarTab to="/infection">Infection</AppBarTab>
                    <AppBarTab to="/player">Player</AppBarTab>
                </AppBarNav>
            </AppBar>
        );
    }
}
