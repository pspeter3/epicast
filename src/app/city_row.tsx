import * as React from "react";
import { FormField } from "../theme/form_field";

export interface Props {
    city: string;
    value: number;
    onChange: (city: string, value: number) => void;
}

export class CityRow extends React.PureComponent<Props, {}> {
    public static displayName = "CityRow";

    public render() {
        const { city, value } = this.props;
        return <FormField label={city} value={value} onChange={this._onChange} />;
    }

    private _onChange = (value: number): void => {
        this.props.onChange(this.props.city, value);
    };
}
