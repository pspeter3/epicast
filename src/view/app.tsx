import * as React from "react";
import { Route } from "react-router-dom";
import { IconButton } from "../theme/buttons";
import { RotateCwwIcon, SettingsIcon } from "../theme/icons";
import { AppBar, Row } from "../theme/layout";
import { IconLink, TabLink } from "../theme/links";
import { Title } from "../theme/typography";
import { Dashboard } from "./dashboard";
import { Discard } from "./discard";
import { Infection } from "./infection";
import { Settings } from "./settings";

export const App: React.SFC<{}> = () => {
    return (
        <React.Fragment>
            <AppBar>
                <Row>
                    <Title>Epidemia</Title>
                    <IconButton>
                        <RotateCwwIcon />
                    </IconButton>
                    <IconLink to="/settings">
                        <SettingsIcon />
                    </IconLink>
                </Row>
                <Row>
                    <TabLink to="/">Dashboard</TabLink>
                    <TabLink to="/infection">Infection</TabLink>
                    <TabLink to="/discard">Discard</TabLink>
                </Row>
            </AppBar>
            <Route path="/" exact component={Dashboard} />
            <Route path="/infection" exact component={Infection} />
            <Route path="/discard" exact component={Discard} />
            <Route path="/settings" exact component={Settings} />
        </React.Fragment>
    );
};
