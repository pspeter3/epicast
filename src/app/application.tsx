import * as React from "react";
import { infectionRate } from "../core/selectors";
import { Config, Game, Stack, State } from "../core/types";
import { configure, epidemic, infect, undo, update } from "../core/updaters";
import { DialogService, StorageService } from "../util/services";
import { size } from "../util/stacks";
import { Dashboard } from "./dashboard";
import { Debug } from "./debug";
import { Infect } from "./infect";
import { Routes } from "./routes";
import { Settings } from "./settings";

export interface Services {
    dialog: DialogService;
    storage: StorageService;
}

export interface Props {
    location: string;
    namespace: string;
    services: Services;
}

export class Application extends React.PureComponent<Props, State> {
    public static displayName = "Application";

    private static _currentGame(state: State): Game {
        return state.games[state.games.length - 1];
    }

    public state = this._initialState();

    public render() {
        const { location } = this.props;
        if (this.state.games.length === 0 || location === Routes.Settings) {
            return (
                <Settings
                    config={this.state.config}
                    onConfigure={this._onConfigure}
                    services={this.props.services}
                />
            );
        }
        const game = Application._currentGame(this.state);
        switch (location) {
            case Routes.Infect:
                return <Infect infection={game.infection} onInfect={this._onInfect} />;
            case Routes.Debug:
                return <Debug state={this.state} />;
        }
        return <Dashboard game={game} onEpidemic={this._onEpidemic} onUndo={this._onUndo} />;
    }

    private _initialState(): State {
        const data = this.props.services.storage.getItem(this.props.namespace);
        return data !== null
            ? JSON.parse(data)
            : {
                  config: {
                      cards: 0,
                      epidemics: 0,
                      cities: {},
                  },
                  games: [],
              };
    }

    private _onConfigure = (config: Config): boolean => {
        if (config.cards <= 0) {
            return this._alert(`Cannot create a game with ${config.cards} in the player deck.`);
        }
        if (config.epidemics <= 0) {
            return this._alert(`Cannot create a game with ${config.epidemics} epidemics.`);
        }
        if (config.cards < config.epidemics) {
            return this._alert(
                `Cannot create a game with fewer cards in the player deck than epidemics.`,
            );
        }
        if (size(config.cities) === 0) {
            return this._alert(`Cannot create a game with no infection cards.`);
        }
        const response = this.props.services.dialog.confirm("Reset game state?");
        if (!response) {
            return response;
        }
        this.setState({ config, games: [configure(config)] });
        return true;
    };

    private _onEpidemic = (city: string) => {
        this.setState(update(this.state, epidemic(Application._currentGame(this.state), city)));
    };

    private _onInfect = (cities: Stack): boolean => {
        const game = Application._currentGame(this.state);
        const rate = infectionRate(game);
        const cards = size(cities);
        if (cards !== 0 && cards !== infectionRate(game)) {
            return this._alert(
                `Cannot infect with ${cards} ${
                    cards === 1 ? "city" : "cities"
                }. Infection rate is ${rate}.`,
            );
        }
        this.setState(update(this.state, infect(game, cities)));
        return true;
    };

    private _onUndo = () => {
        if (this.state.games.length < 2) {
            this._alert("Cannot undo any more actions.");
            return;
        }
        this.setState(undo);
    };

    private _alert(msg: string): boolean {
        this.props.services.dialog.alert(msg);
        return false;
    }
}
