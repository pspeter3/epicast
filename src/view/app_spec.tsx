import * as React from "react";
import { MemoryRouter } from "react-router";
import { snapshotSuite } from "../util/snapshot";
import { App } from "./app";

snapshotSuite("App", () => (
    <MemoryRouter>
        <App />
    </MemoryRouter>
));
