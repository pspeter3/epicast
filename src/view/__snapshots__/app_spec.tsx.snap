// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
Array [
  <header
    className="appbar"
  >
    <div
      className="row"
    >
      <h1
        className="title"
      >
        Epidemia
      </h1>
      <button
        className="icon-button"
      >
        <svg
          className="feather feather-rotate-ccw"
          fill="none"
          height="24"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="1 4 1 10 7 10"
          />
          <path
            d="M3.51 15a9 9 0 1 0 2.13-9.36L1 10"
          />
        </svg>
      </button>
      <a
        aria-current={false}
        className="icon-link"
        href="/settings"
        onClick={[Function]}
        style={undefined}
      >
        <svg
          className="feather feather-settings"
          fill="none"
          height="24"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="3"
          />
          <path
            d="M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z"
          />
        </svg>
      </a>
    </div>
    <div
      className="row"
    >
      <a
        aria-current="true"
        className="tab-link active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        Dashboard
      </a>
      <a
        aria-current={false}
        className="tab-link"
        href="/infection"
        onClick={[Function]}
        style={undefined}
      >
        Infection
      </a>
      <a
        aria-current={false}
        className="tab-link"
        href="/discard"
        onClick={[Function]}
        style={undefined}
      >
        Discard
      </a>
    </div>
  </header>,
  <section
    className="section"
  >
    <div
      className="row"
    >
      <h2
        className="header"
      >
        Epidemic Probability
      </h2>
      <p
        className="data-percentage"
      >
        63
        %
      </p>
    </div>
    <div
      className="row"
    >
      <meter
        className="data-meter active"
        value={0.63}
      />
    </div>
  </section>,
  <main
    className="main"
  >
    <div
      className="row"
    >
      <p
        className="text"
      >
        Table Goes Here
      </p>
    </div>
  </main>,
]
`;
